vcd/resource_vcd_nsxt_edgegateway.go

```diff
@@ -0,0 +1,414 @@
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error creating NSX-T edge gateway: %s", err))
+	}
+
+	d.SetId(createdEdgeGateway.EdgeGateway.ID)
+
+	return resourceVcdNsxtEdgeGatewayRead(ctx, d, meta)
+}
+
+// resourceVcdNsxtEdgeGatewayUpdate
+func resourceVcdNsxtEdgeGatewayUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[TRACE] NSX-T edge gateway update initiated")
+
+	vcdClient := meta.(*VCDClient)
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error retrieving VDC: %s", err))
+	}
+
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error getting adminOrg: %s", err))
+	}
+
+	edge, err := adminOrg.GetNsxtEdgeGatewayById(d.Id())
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("could not retrieve NSX-T edge gateway: %s", err))
+	}
+
+	updatedEdge, err := getNsxtEdgeGatewayType(d, vdc)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error updating NSX-T edge gateway type: %s", err))
+	}
+
+	updatedEdge.ID = edge.EdgeGateway.ID
+	edge.EdgeGateway = updatedEdge
+
+	_, err = edge.Update(edge.EdgeGateway)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error updating edge gateway with ID '%s': %s", d.Id(), err))
+	}
+
+	return resourceVcdNsxtEdgeGatewayRead(ctx, d, meta)
+}
+
+// resourceVcdNsxtEdgeGatewayRead
+func resourceVcdNsxtEdgeGatewayRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[TRACE] NSX-T edge gateway read initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error retrieving VDC: %s", err))
+	}
+
+	edge, err := vdc.GetNsxtEdgeGatewayById(d.Id())
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("could not retrieve NSX-T edge gateway: %s", err))
+	}
+
+	err = setNsxtEdgeGatewayData(edge.EdgeGateway, d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading NSX-T edge gateway data: %s", err))
+	}
+	return nil
+}
+
+// resourceVcdNsxtEdgeGatewayDelete
+func resourceVcdNsxtEdgeGatewayDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[TRACE] edge gateway deletion initiated")
+
+	vcdClient := meta.(*VCDClient)
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error retrieving VDC: %s", err))
+	}
+
+	edge, err := vdc.GetNsxtEdgeGatewayById(d.Id())
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("could not retrieve NSX-T edge gateway: %s", err))
+	}
+
+	err = edge.Delete()
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error deleting NSX-T edge gateway: %s", err))
+	}
+
+	return nil
+}
+
+// resourceVcdNsxtEdgeGatewayImport
+func resourceVcdNsxtEdgeGatewayImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```

Comment:
if user would need list and import by ID functionality?

Reply:
Hm, we used to provide list and import when name could wasn't required to be unique. In this case edge gateway names must be unique. No problems to add this if we agree. (Although now I think that we could even refer to `vcd_resource_list` datasource for listing existing resources.)

Reply:
if we can use vcd_resource_list for most of the things, maybe is worth it to create issue to remove custom listings from curent resources? @dataclouder

Reply:
I think the current listings in import functions have their limited usability, i.e. mostly provide advice on duplicate names.
Removing them would make the functions less usable without providing any immediate benefit.

Reply:
> No problems to add this if we agree. (Although now I think that we could even refer to vcd_resource_list datasource for listing existing resources.)

Maybe let's add that reference in the docs here? Maybe even a small example (just a thought)? Otherwise, it's not clear that it's possible and useful here, esp. given that `vcd_resource_list` is fresh.

Reply:
Added a note and reference to `vcd_resource_list`